### ISSUE TITLE
Termios addon to allow for more typical terminal behavior 

### ISSUE THREAD

**Issue by upsampled** [xtermjs/xterm.js#3394]
I agree with the maintainers regarding #2664 , a full PTY implementation is out-of-scope and is more involved than one would initially realize; however, I believe just a 'termios' implementation would be a huge benefit for the majority of developers and fall within the projects goals.  

It's pretty common for developer, me at least, to throw a pty infront of raw application data to get sensical terminal behavior. Most immediately it handles the CR, NL, and CRNL behavior that is seen in a typical shell environment. In these cases I am not really using the pty for anything other than [termios ](https://man7.org/linux/man-pages/man3/termios.3.html) flags. These flags are also used by essential, terminal aware cli tools to perform tasks like [reading passwords ](https://github.com/openssh/openssh-portable/blob/2dc328023f60212cd29504fc05d849133ae47355/openbsd-compat/readpassphrase.c#L90).

Right now it looks like many developers in order to get typical typical terminal behavior are:
* piping a pty to/from a host 
* creating custom `onData`/`onmessage` functions to implement the behavior they want. 

I believe that a termios addon would:
* eliminate the need to pipe a pty in many cases
* allow 'normal' terminal behavior without custom data handlers
  * Just pass they typical termios flags to the addon for normal behavior 
* set a better foundation for porting common cli applications
  * Providing a `tcgetattr`, `tcsetattr`, and `cfmakeraw` callback to WASM code should handle terminal aware applications

I am just about to refactor some custom ondata methods to take a termios argument, so i figured I'd ask if it was worth making into a full addon first.

**Comment by jerch** [comment]
I started investigating into that direction in https://github.com/jerch/browser-fakepty. Basic things are working, but it still has loose ends (mostly around proper backpressure/flow control handling and process control shims). I started to shim some process API, see https://github.com/jerch/browser-fakepty/blob/master/src/Shell.ts for example "processes" and a naive shell impl. Note that the code is not even alpha, it is just a testground about possible directions with a fake pty.

Maybe you find there some ideas, how to go about the missing-pty issue..

**Comment by Kreijstal** [comment]
soooo how is this going?

**Comment by RReverser** [comment]
> set a better foundation for porting common cli applications
> * Providing a `tcgetattr`, `tcsetattr`, and `cfmakeraw` callback to WASM code should handle terminal aware applications

For those coming across this issue specifically for Wasm usecase - [xterm-pty](https://github.com/mame/xterm-pty) solves this for Emscripten nowadays, covering line discipline, terminal sizing, and, with [an upcoming PR](https://github.com/mame/xterm-pty/pull/10) converting things like Ctrl+C to signals as well.

**Comment by jerch** [comment]
@RReverser Oh interesting - thx for the headsup. May I ask how you are going to solve the signals issue? When I was looking into this, it kinda pulled half of the POSIX landscape into it including process management, which I found rather inconvenient, as JS code execution is not interuptable. Only termination with the help of webworkers was doable.
Do you hook in some kernel-like functionality?

**Comment by RReverser** [comment]
> May I ask how you are going to solve the signals issue?

Emscripten already had basic intra-process signal emulation support except for default signal handlers, which I added too. https://github.com/emscripten-core/emscripten/pull/20257. WASI SDK does the same.

Of course, inter-process signals won't work, but this is enough for being able to trigger signal via `libc::raise` from one place (e.g. terminal keystroke handler in our case) and handle in another, which covers basic stuff like termination, terminal resizing and such.

> as JS code execution is not interuptable (which yet again boils down to real blocking semantics not being available in JS/WASM)

There are couple of ways to do it, you can see PR description and docs for more details as it was pretty much full Emscripten integration rewrite rather than just signals. Basically it supports two modes in order to suspend Wasm and read from keyboard or poll - one is Asyncify which allows async JS APIs to be called while suspending entire Wasm and another is PROXY_TO_PTHREAD which transparently moves the entire Wasm to separate thread so can have same effect (block execution thread of Wasm, but not I/O that's happening on the main thread).

**Comment by jerch** [comment]
Yeah blocking "syscalls" should be doable with asyncify on the mainthread. And in a worker the asyncify overhead could be avoided by using atomic wait conditions. Still thats only working for actions the code asked for in the first place like reading from a blocking source.

But a POSIX signal can interrupt execution at any time, and would run the signal handler instead. How would that unconditional execution jump in middle of anywhere work with WASM? How would that be triggered and where would it jump to? Does WASI shim some magic kernel-like interface with these superpowers into the binary? Last time I read about WASI it was more around IO, but not about unsolicited execution traps orchestrated from outside of the code (kernel like). Ofc there is a chance that I got WASI concepts partially wrong.

**Comment by jerch** [comment]
> Anyway, this is probably not the right place for discussion

Yeah sorry, your are right. Thx for the pointer :smiley_cat:

**Comment by Tyriar** [comment]
This is out of scope for this project, https://github.com/jerch/browser-fakepty is the closest thing afaik
